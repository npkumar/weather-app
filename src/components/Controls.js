import React, { useState } from "react";

import {
  FormControl,
  RadioGroup,
  FormControlLabel,
  Radio,
} from "@material-ui/core";
import {
  useApplication,
  FAHRENHEIT,
  CELSIUS,
  KELVIN,
} from "../services/Application";

const Controls = () => {
  const { application, setUnit } = useApplication();
  const [value, setValue] = useState(application.unit);
  const handleChange = (e) => {
    setValue(e.target.value);
    setUnit(e.target.value);
  };

  return (
    <FormControl component="fieldset">
      <RadioGroup
        row
        aria-label="temperature"
        name="temperature"
        value={value}
        onChange={handleChange}
        data-testid="radioGroup"
      >
        <FormControlLabel
          value={FAHRENHEIT}
          control={<Radio />}
          label="Fahrenheit"
          data-testid="fahrenheit"
        />
        <FormControlLabel
          value={CELSIUS}
          control={<Radio />}
          label="Celsius"
          data-testid="celsius"
        />
        <FormControlLabel
          value={KELVIN}
          control={<Radio />}
          label="Kelvin"
          data-testid="kelvin"
        />
      </RadioGroup>
    </FormControl>
  );
};

export default Controls;
