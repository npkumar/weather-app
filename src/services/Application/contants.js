export const FAHRENHEIT = "fahrenheit";
export const KELVIN = "kelvin";
export const CELSIUS = "celsius";
