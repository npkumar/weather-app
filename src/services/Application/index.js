// Hooks
export { default as useApplication } from "./useApplication";

// Providers
export { default as ApplicationProvider } from "./ApplicationProvider";

// Constants
export { FAHRENHEIT, CELSIUS, KELVIN } from "./contants";
