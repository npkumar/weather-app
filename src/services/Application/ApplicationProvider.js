import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";
import * as constants from "./contants";

// Actions
const SET_UNIT = "SET_UNIT";
const SET_INDEX = "SET_INDEX";

export const ApplicationContext = createContext(null);

const ApplicationProvider = ({ children }) => {
  const initialState = {
    unit: constants.FAHRENHEIT,
    index: 0,
  };

  function reducer(state, action) {
    switch (action.type) {
      case SET_UNIT: {
        return { ...state, unit: action.payload };
      }
      case SET_INDEX: {
        return { ...state, index: action.payload };
      }
      default:
        throw new Error();
    }
  }

  const [state, dispatch] = useReducer(reducer, initialState);

  const value = {
    application: state,
    setUnit: (unit) => dispatch({ type: SET_UNIT, payload: unit }),
    setIndex: (index) => dispatch({ type: SET_INDEX, payload: index }),
  };

  return (
    <ApplicationContext.Provider value={value}>
      {children}
    </ApplicationContext.Provider>
  );
};

ApplicationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ApplicationProvider;
