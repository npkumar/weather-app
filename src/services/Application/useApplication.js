import { useContext } from "react";
import { ApplicationContext } from "./ApplicationProvider";

const useApplication = () => useContext(ApplicationContext);

export default useApplication;
