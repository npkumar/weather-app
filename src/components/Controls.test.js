import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import Controls from "./Controls";
import { ApplicationContext } from "../services/Application/ApplicationProvider";
import * as constants from "../services/Application/contants";

const value = {
  application: {
    unit: constants.FAHRENHEIT,
  },
};

describe("Controls", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { asFragment } = render(
      <ApplicationContext.Provider value={value}>
        <Controls />
      </ApplicationContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have 3 radio buttons", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider value={value}>
        <Controls />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("radioGroup")).toBeInTheDocument();
    expect(getByTestId("fahrenheit")).toHaveTextContent("Fahrenheit");
    expect(getByTestId("celsius")).toHaveTextContent("Celsius");
    expect(getByTestId("kelvin")).toHaveTextContent("Kelvin");
  });

  it("fahrenheit option should be checked by default", () => {
    const { getByDisplayValue } = render(
      <ApplicationContext.Provider value={value}>
        <Controls />
      </ApplicationContext.Provider>
    );
    expect(getByDisplayValue("fahrenheit")).toBeChecked();
  });
});
