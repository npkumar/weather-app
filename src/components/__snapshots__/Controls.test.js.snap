// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls renders 1`] = `
<DocumentFragment>
  <fieldset
    class="MuiFormControl-root"
  >
    <div
      aria-label="temperature"
      class="MuiFormGroup-root MuiFormGroup-row"
      data-testid="radioGroup"
      role="radiogroup"
    >
      <label
        class="MuiFormControlLabel-root"
        data-testid="fahrenheit"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input-4"
              name="temperature"
              type="radio"
              value="fahrenheit"
            />
            <div
              class="PrivateRadioButtonIcon-root-5 PrivateRadioButtonIcon-checked-7"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Fahrenheit
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
        data-testid="celsius"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-4"
              name="temperature"
              type="radio"
              value="celsius"
            />
            <div
              class="PrivateRadioButtonIcon-root-5"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Celsius
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
        data-testid="kelvin"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-4"
              name="temperature"
              type="radio"
              value="kelvin"
            />
            <div
              class="PrivateRadioButtonIcon-root-5"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Kelvin
        </span>
      </label>
    </div>
  </fieldset>
</DocumentFragment>
`;
